Public/Src/Sandbox/MacOs/Sandbox/Src/SandboxedPip.cpp

```diff
@@ -88,12 +89,46 @@ PipInfo SandboxedPip::introspect() const
         .pid                 = getProcessId(),
         .clientPid           = getClientPid(),
         .pipId               = getPipId(),
-        .cacheSize           = pathCache_->getCount(),
+        .cacheSize           = getPathCacheElemCount(),
         .treeSize            = getTreeSize(),
         .counters            = counters_,
         .numReportedChildren = 0,
         .children            = {0}
     };
 }
 
+bool SandboxedPip::RefreshDisableCaching()
+{
+    if (!disableCaching_)
+    {
+        if (ShouldDisableCaching())
+        {
+            disableCaching_ = true;
+            Trie *oldCache = pathCache_;
+            Trie *newCache = Trie::createPathTrie();
+            if (OSCompareAndSwapPtr(oldCache, newCache, &pathCache_))
+            {
+                // we swapped --> release oldCache
+                OSSafeReleaseNULL(oldCache);
+            }
+            else
+            {
+                // someone else did it --> release newCache
+                OSSafeReleaseNULL(newCache);
+            }
+        }
+    }
+
+    return disableCaching_;
+}
+
+inline bool SandboxedPip::ShouldDisableCaching()
+{
+    return
+        // have more than 30000 cache entries
+        pathCache_->getCount() > 20000 &&
```

Comment:
Are these heuristics specific to the use case we are currently investigating or do you think they are in general cut off numbers that are safe and sound?

Reply:
made it configurable